{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Required packages and softwares\n",
    "Install required softwares with conda and python packages with pip\n",
    "```bash\n",
    "conda install -c bioconda samtools\n",
    "conda install -c bioconda seqtk\n",
    "conda install -c bioconda blast\n",
    "conda install -c bioconda mappy\n",
    "conda install -c bioconda tidehunter\n",
    "pip install numpy\n",
    "pip install pandas\n",
    "pip install biopython\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Structure of folders in this pipeline\n",
    "```bash\n",
    "|-- ref/ #reference genome assembly\n",
    "|    |-- c_elegans.PRJNA13758.WS270.genomic.fa #C. elegans ce11 genome assembly\n",
    "|    |-- c_elegans.PRJNA13758.WS270.genomic.fa.fai #ce11 index file\n",
    "|-- chr_ref/ #extracted genome sequences for blastn against\n",
    "|    |-- N2_IL.fa\n",
    "|    |-- ...\n",
    "|-- blastn/ #blastn results\n",
    "|    |-- N2EM_IL.tsv\n",
    "|    |-- ...\n",
    "|-- fasta/ #extracted read sequences from BAM file\n",
    "|    |-- N2EM_IL.fa\n",
    "|    |-- N2EM_IL.fa.fai\n",
    "|    |-- ...\n",
    "|-- read_list_blastn/ #read name list from blastn result TSV files\n",
    "|    |-- N2EM_IL_list\n",
    "|    |-- ...\n",
    "|-- trimming_result/ #folders to save statistic results and plot\n",
    "|    |-- log/ #statistic results of trimming information and telomeric sequences containing reads\n",
    "|        |-- N2_telomere_stats.tsv\n",
    "|        |-- N2_trimming_result.tsv\n",
    "|    |-- ONT_reads/ #full-length nanopore read sequences\n",
    "|        |-- N2EM_IR_F.fa\n",
    "|        |-- N2EM_IR_F.fa.fail\n",
    "|        |-- N2EM_IL_F_list\n",
    "|        |-- ...\n",
    "|    |-- strand/ #telomere sequences listed by stage, chromosome end,and strand\n",
    "|        |-- N2EM_IR_F.fa\n",
    "|        |-- N2EM_IR_F.fa.fail\n",
    "|        |-- ...\n",
    "|    |-- telomere_only/ #telomere sequences listed by chromosome end\n",
    "|        |-- N2EM_IL.fa\n",
    "|        |-- N2EM_IL.fa.fai\n",
    "|        |-- N2EM_IL_list\n",
    "|    |-- tidehunter/ #TideHunter results and telomere repeat motif results including copy number and coverage information\n",
    "|        |-- stats/ #for ggplot2 plotting\n",
    "|            |-- coverage/ #For plotting strand dependent telomere repeat motif abundance\n",
    "|                |-- N2EM_IL_F_Coverage.tsv\n",
    "|                |-- ...\n",
    "|            |-- unsort/\n",
    "|                |-- N2EM_IL_F_stats.tsv\n",
    "|                |-- ...\n",
    "|            |-- sorted/ #For converting telomere repeat motif in R\n",
    "|                |-- N2EM_IL_F_stats_sorted.tsv\n",
    "|                |-- ...            \n",
    "|        |-- N2EM_IR_F_P8p4k8.tsv #TideHunter tabular result for each file\n",
    "|        |-- ...\n",
    "|        |-- N2YA_XR_R_P8p4k8.tsv\n",
    "|    |-- plot_file/ #Plotting files \n",
    "|        |-- motif_proportion/ #Motif coverage plots by stage, chromosome end, and strand\n",
    "|            |-- N2EM_IL_F_proportion_6X6.pdf\n",
    "|            |-- N2EM_IL_R_proportion_6X6.pdf\n",
    "|            |-- ...\n",
    "|        |-- motif_cluster_read/ #Telomere repeat motif in reads\n",
    "|            |-- N2EM_IL_F_P8p4k3_6X6.pdf\n",
    "|            |-- ...\n",
    "|        |-- ONT_reads_with_telomere_length_strand_table.csv #Full-length nanopore reads information\n",
    "|        |-- telomere_length_strand_table.csv #Telomere sequences listed by stage, chromosome end, and strand\n",
    "|        |-- telomere_length_table.csv #Telomere sequences listed by stage and chromosome end\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import required packages and functions\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "import mappy as mp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#change to working directory \n",
    "os.chdir('/mnt/f/nanopore_data/telomere/N2_extract')\n",
    "#reference genome should be put under this directory, i.e.  ../N2_extract/ref/genome.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get chromosome end reference from genome assembly (C. elegans genome assembly ce11 or WBcel235)\n",
    "#genome assembly can be download in https://wormbase.org/species/all#1--10\n",
    "genome_file = str(\"ref/c_elegans.PRJNA13758.WS270.genomic.fa\")\n",
    "#load genome assembly for determine chimeric sequence from ONT reads\n",
    "cel_genome = mp.Aligner(genome_file, preset=\"map-ont\")\n",
    "if not cel_genome: raise Exception(\"ERROR: failed to load/build index\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare list for loopcommand\n",
    "stg = [\"N2EM\",\"N2L1\",\"N2YA\"]\n",
    "chromosome = [\"I\", \"II\", \"III\", \"IV\", \"V\", \"X\"]\n",
    "chrend = [\"L\", \"R\"]\n",
    "strand = [\"F\", \"R\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make directory if not exist\n",
    "!mkdir -p chr_ref/#reference sequences\n",
    "!mkdir -p blastn/#save blastn results\n",
    "!mkdir -p read_list_blastn/#filtered read list\n",
    "!mkdir -p trimming_result/#directory to save trimmed telomere sequences\n",
    "!mkdir -p trimming_result/log/#directory to save statistic result \n",
    "!mkdir -p trimming_result/strand/ #telomere sequences by strand at each chromosome end\n",
    "!mkdir -p trimming_result/telomere_only/ #telomere sequences by chromosome end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#index genome assembly by samtools\n",
    "!samtools faidx $genome_file\n",
    "#extract chromosome end sequence and convert one direction\n",
    "!samtools faidx $genome_file \"I:433-10433\" |seqtk seq -r > chr_ref/N2_IL.fa\n",
    "!samtools faidx $genome_file \"I:15068359-15069276\" |seqtk seq > chr_ref/N2_IR.fa\n",
    "!samtools faidx $genome_file \"II:174-10174\" |seqtk seq -r > chr_ref/N2_IIL.fa\n",
    "!samtools faidx $genome_file \"II:15269231-15279231\" |seqtk seq > chr_ref/N2_IIR.fa\n",
    "!samtools faidx $genome_file \"III:132-10132\" |seqtk seq -r > chr_ref/N2_IIIL.fa\n",
    "!samtools faidx $genome_file \"III:13773443-13783443\" |seqtk seq > chr_ref/N2_IIIR.fa\n",
    "!samtools faidx $genome_file \"IV:150-10150\" |seqtk seq -r > chr_ref/N2_IVL.fa\n",
    "!samtools faidx $genome_file \"IV:17483620-17493620\" |seqtk seq > chr_ref/N2_IVR.fa\n",
    "!samtools faidx $genome_file \"V:328-10328\" |seqtk seq -r > chr_ref/N2_VL.fa\n",
    "!samtools faidx $genome_file \"V:20913997-20923997\" |seqtk seq > chr_ref/N2_VR.fa\n",
    "!samtools faidx $genome_file \"X:18393-22393\" |seqtk seq > chr_ref/N2_XL.fa\n",
    "!samtools faidx $genome_file \"X:17708805-17718805\" |seqtk seq > chr_ref/N2_XR.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#the extracted chromosome end reads were blastn against chromosome end sequences\n",
    "#generate a file of read name list with bash command cut and awk to extract the read name from blastn result(the 2nd column)\n",
    "for stage in stg:\n",
    "    for chrom in chromosome:\n",
    "        for position in chrend:\n",
    "            #input file path\n",
    "            read_file = str(\"fasta/%s_%s%s.fa\" % (stage, chrom, position))\n",
    "            ref_file = str(\"chr_ref/N2_%s%s.fa\" % (chrom, position))\n",
    "            #output blastn result and read name list\n",
    "            blastn_file = str(\"blastn/%s_%s%s.tsv\" % (stage, chrom, position))\n",
    "            read_list_file = str(\"read_list_blastn/%s_%s%s_list\" % (stage, chrom, position))\n",
    "            #index extracted input fasta file\n",
    "            !samtools faidx $read_file\n",
    "            #against chromosome end sequence to extracted reads\n",
    "            !blastn -query $ref_file -subject $read_file -outfmt \"6 std sstrand\" -max_target_seqs 999999999 -out $blastn_file\n",
    "            #get the read name list from blastn result\n",
    "            !cut -f2 $blastn_file | awk '!seen[$0]++' > $read_list_file\n",
    "            print(str(\"%s_%s%s finished\" % (stage, chrom, position)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#nanopore reads sequence filtering parameter\n",
    "#align trimmed sequences to the C. elegans genome assembly and check whether it is chimeric\n",
    "def checkChimeric(trimmed_seq):\n",
    "    chimeric = False\n",
    "    try:\n",
    "        for hit in cel_genome.map(trimmed_seq):\n",
    "            if hit.mapq == 60:\n",
    "                chimeric = True\n",
    "        return(chimeric)\n",
    "    except:\n",
    "        return(chimeric)\n",
    "def IR_isEnd(trimmed_seq):\n",
    "    SeqisEnd = True\n",
    "    try:\n",
    "        for hit in cel_genome.map(trimmed_seq):\n",
    "            if hit.ctg == \"I\" and hit.r_st > 15062000 and hit.r_en < 15072265 and hit.blen > 200:\n",
    "                SeqisEnd = False\n",
    "        return(SeqisEnd)\n",
    "    except:\n",
    "        return(SeqisEnd)\n",
    "#trim reads by blastn against chromosome end sequences\n",
    "def trimSeq(df_blast):\n",
    "    global chrom, position, forward_counter, reverse_counter, short_counter, non_end_counter,\\\n",
    "    self_duplication_counter, chimeric_counter, read_total_length, telomere_total_length, \\\n",
    "    readName, forward_total_length, reverse_total_length\n",
    "    if chrom == \"I\" and position == \"R\":\n",
    "        if len(df_blast) == 0:#non-chromosomal end reads\n",
    "            non_end_counter += 1\n",
    "        elif len(df_blast) == 2:#self duplication reads\n",
    "            self_duplication_counter += 1\n",
    "        else:\n",
    "            if  str(df_blast.iloc[0,-1]) == \"plus\":#if read belongs to forward strand\n",
    "                telomere_end = df_blast[\"send\"].max()\n",
    "                out_seq = read_seq[telomere_end:]\n",
    "                if len(out_seq) < 12:#skip too short telomere sequence < 12 bp\n",
    "                    short_counter += 1\n",
    "                    read_total_length += len(read_seq)\n",
    "                else:\n",
    "                    if checkChimeric(out_seq) == True:#if trimmed sequences have other chomosomal alignment\n",
    "                        chimeric_counter +=1\n",
    "                    elif IR_isEnd(out_seq) == False:\n",
    "                        non_end_counter += 1\n",
    "                    else:\n",
    "                        forward_out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        forward_out_w.write(out_seq+\"\\n\")\n",
    "                        out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        out_w.write(out_seq+\"\\n\")\n",
    "                        forward_counter += 1\n",
    "                        forward_total_length += len(out_seq)\n",
    "                        read_total_length += len(read_seq)\n",
    "                        telomere_total_length += len(out_seq)\n",
    "            else:#if read belongs to reverse strand\n",
    "                telomere_end = df_blast[\"send\"].min()\n",
    "                out_seq = Seq.reverse_complement(read_seq[:telomere_end-1])\n",
    "                if len(out_seq) < 12:#skip too short telomere length < 12 bp\n",
    "                    short_counter += 1\n",
    "                    read_total_length += len(read_seq)\n",
    "                else:\n",
    "                    if checkChimeric(out_seq) == True:#if trimmed sequences have other chomosomal alignment\n",
    "                        chimeric_counter +=1\n",
    "                    elif IR_isEnd(out_seq) == False:\n",
    "                        non_end_counter += 1\n",
    "                    else:\n",
    "                        reverse_out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        reverse_out_w.write(out_seq+\"\\n\")\n",
    "                        out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        out_w.write(out_seq+\"\\n\")\n",
    "                        reverse_counter += 1\n",
    "                        reverse_total_length += len(out_seq)\n",
    "                        read_total_length += len(read_seq)\n",
    "                        telomere_total_length += len(out_seq)\n",
    "    elif chrom == \"IV\" and position == \"R\":\n",
    "        df_blast_process = df_blast[df_blast[\"length\"] > 2000]\n",
    "        if len(df_blast) == 0:#non-chromosomal end reads\n",
    "            non_end_counter += 1\n",
    "        elif len(df_blast_process) > 1:#self duplication reads\n",
    "            self_duplication_counter += 1\n",
    "        else:\n",
    "            if  df_blast.loc[[df_blast[\"length\"].idxmax()]].iloc[0,-1] == \"plus\":#if read belongs to forward strand\n",
    "                out_seq = read_seq[df_blast[\"send\"].max():]\n",
    "                if len(out_seq) < 12:#skip too short telomere sequence < 12 bp\n",
    "                    short_counter += 1\n",
    "                    read_total_length += len(read_seq)\n",
    "                else:\n",
    "                    if checkChimeric(out_seq) == True:#if trimmed sequences have alignment\n",
    "                        chimeric_counter +=1\n",
    "                    else:\n",
    "                        forward_out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        forward_out_w.write(out_seq+\"\\n\")\n",
    "                        out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        out_w.write(out_seq+\"\\n\")\n",
    "                        forward_counter += 1\n",
    "                        forward_total_length += len(out_seq)\n",
    "                        read_total_length += len(read_seq)\n",
    "                        telomere_total_length += len(out_seq)\n",
    "            else:#if read belongs to reverse strand\n",
    "                out_seq = Seq.reverse_complement(read_seq[:df_blast[\"send\"].min() - 1])\n",
    "                if len(out_seq) < 12:#skip too short telomere length < 12 bp\n",
    "                    short_counter += 1\n",
    "                    read_total_length += len(read_seq)\n",
    "                else:\n",
    "                    if checkChimeric(out_seq) == True:#if trimmed sequences have other chomosomal alignment\n",
    "                        chimeric_counter +=1\n",
    "                    else:\n",
    "                        reverse_out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        reverse_out_w.write(out_seq+\"\\n\")\n",
    "                        out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        out_w.write(out_seq+\"\\n\")\n",
    "                        reverse_counter += 1\n",
    "                        reverse_total_length += len(out_seq)\n",
    "                        read_total_length += len(read_seq)\n",
    "                        telomere_total_length += len(out_seq)\n",
    "    else: \n",
    "        if len(df_blast) == 0:#non-chromosomal end reads\n",
    "            non_end_counter += 1\n",
    "        elif len(df_blast) == 2:#self duplication reads\n",
    "            self_duplication_counter += 1\n",
    "        else:\n",
    "            if  str(df_blast.iloc[0,-1]) == \"plus\":#if read belongs to forward strand\n",
    "                telomere_end = df_blast[\"send\"].max()\n",
    "                out_seq = read_seq[telomere_end:]\n",
    "                if len(out_seq) < 12:#skip too short telomere sequence < 12 bp\n",
    "                    short_counter += 1\n",
    "                    read_total_length += len(read_seq)\n",
    "                else:\n",
    "                    if checkChimeric(out_seq) == True:#if trimmed sequences have other chomosomal alignment\n",
    "                        chimeric_counter +=1\n",
    "                    else:\n",
    "                        forward_out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        forward_out_w.write(out_seq+\"\\n\")\n",
    "                        out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        out_w.write(out_seq+\"\\n\")\n",
    "                        forward_counter += 1                        \n",
    "                        forward_total_length += len(out_seq)\n",
    "                        read_total_length += len(read_seq)\n",
    "                        telomere_total_length += len(out_seq)\n",
    "            else:#if read belongs to reverse strand\n",
    "                telomere_end = df_blast[\"send\"].min()\n",
    "                out_seq = Seq.reverse_complement(read_seq[:telomere_end-1])\n",
    "                if len(out_seq) < 12:#skip too short telomere length < 12 bp\n",
    "                    short_counter += 1\n",
    "                    read_total_length += len(read_seq)\n",
    "                else:\n",
    "                    if checkChimeric(out_seq) == True:#if trimmed sequences have other chomosomal alignment\n",
    "                        chimeric_counter +=1\n",
    "                    else:\n",
    "                        reverse_out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        reverse_out_w.write(out_seq+\"\\n\")\n",
    "                        out_w.write(str(\">%s\\n\" %(readName)))\n",
    "                        out_w.write(out_seq+\"\\n\")\n",
    "                        reverse_counter += 1\n",
    "                        reverse_total_length += len(out_seq)\n",
    "                        read_total_length += len(read_seq)\n",
    "                        telomere_total_length += len(out_seq)\n",
    "def blastn_filter(chrom, position, readName):\n",
    "    global blastn_info\n",
    "    if chrom == \"I\" and position == \"L\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"length\"] > 200)&(blastn_info[\"qend\"] > 9900)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"I\" and position == \"R\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"length\"] > 200)&(blastn_info[\"qend\"] > 810)&(blastn_info[\"qend\"] < 825)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"II\" and position == \"L\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"length\"] > 500)&(blastn_info[\"qend\"] > 9950)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"II\" and position == \"R\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"length\"] > 500)&(blastn_info[\"qend\"] > 9950)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"III\" and position == \"L\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"qend\"] > 9950)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"III\" and position == \"R\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"length\"] > 200)&(blastn_info[\"qend\"] > 7550)&(blastn_info[\"qend\"] < 7650)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"IV\" and position == \"L\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"length\"] > 200)&(blastn_info[\"qend\"] > 9950)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"IV\" and position == \"R\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"qend\"] > 550)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"V\" and position == \"L\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"length\"] > 300)&(blastn_info[\"qend\"] > 9950)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"V\" and position == \"R\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"qstart\"] < 9050)&(blastn_info[\"qend\"] > 9950)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"X\" and position == \"L\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"qend\"] > 1900)&(blastn_info[\"qend\"] < 2000)]\n",
    "        return (df_return)\n",
    "    elif chrom == \"X\" and position == \"R\":\n",
    "        df_return = blastn_info[(blastn_info[\"subject\"]== readName)&(blastn_info[\"qend\"] > 9950)]\n",
    "        return (df_return)\n",
    "def write_log(stage, chrom, position):\n",
    "    global forward_counter, reverse_counter, short_counter, non_end_counter,\\\n",
    "    self_duplication_counter, chimeric_counter, read_total_length, telomere_total_length,\\\n",
    "    forward_total_length, reverse_total_length\n",
    "    fr_sum = forward_counter + reverse_counter\n",
    "    forward_average_length = forward_total_length / forward_counter\n",
    "    reverse_average_length = reverse_total_length / reverse_counter\n",
    "    read_average_length = read_total_length /(fr_sum + short_counter)\n",
    "    telomere_average_length = telomere_total_length/ (fr_sum)\n",
    "    read_info = str(\"%s\\t%s\\t%s\"%(stage, chrom, position))\n",
    "    log_out.write(\"{}\\t{}\\tTelomere_reads_count\\n\".format(read_info, fr_sum))#Telomere reads count\n",
    "    log_out.write(\"{}\\t{}\\tForward_strand_count\\n\".format(read_info, forward_counter))#Forward strand count\n",
    "    log_out.write(\"{}\\t{:.1f}\\tForward_telomere_average_length\\n\".format(read_info, forward_average_length))\n",
    "    log_out.write(\"{}\\t{:.1f}\\tReverse_strand_count\\n\".format(read_info, reverse_counter))#Reverse strand count\n",
    "    log_out.write(\"{}\\t{:.1f}\\tReverse_telomere_average_length\\n\".format(read_info, reverse_average_length))\n",
    "    log_out.write(\"{}\\t{:.1f}\\tAverage_read_length\\n\".format(read_info, read_average_length))#Average read length\n",
    "    log_out.write(\"{}\\t{:.1f}\\tAverage_telomere_length\\n\".format(read_info, telomere_average_length))#Average telomere length\n",
    "    log_out.write(\"{}\\t{}\\tShort_read_count\\n\".format(read_info, short_counter))#Short read count\n",
    "    log_out.write(\"{}\\t{}\\tNon-end_read_count\\n\".format(read_info, non_end_counter))#Non-end read count\n",
    "    log_out.write(\"{}\\t{}\\tSelf_duplication_count\\n\".format(read_info, self_duplication_counter))#Self duplication count\n",
    "    log_out.write(\"{}\\t{}\\tChimeric_reads_count\\n\".format(read_info, chimeric_counter))#Chimeric reads count\n",
    "    sys.stdout.write('\\r'+\"{}\\tfinished\".format(read_info))\n",
    "def enumerateRepeat(telo_repeat = \"TTAGGC\"):\n",
    "    telo_list = []\n",
    "    for idx, char in enumerate(telo_repeat):\n",
    "        motif = telo_repeat[idx:]+telo_repeat[:idx]\n",
    "        telo_list.append(motif)\n",
    "    return(telo_list)\n",
    "def tide_filter(readName):\n",
    "    global tide_info\n",
    "    df_tide_return = tide_info[(tide_info[\"readName\"]== readName)]\n",
    "    return (df_tide_return)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N2YA\tX\tR\tfinisheddd\n",
      "job finished!\n"
     ]
    }
   ],
   "source": [
    "#file name of statistic results\n",
    "outlog_fh = str(\"trimming_result/log/N2_trimming_result.tsv\")\n",
    "#trim read sequences with blastn result and write statistic result to file\n",
    "#telomere sequences seperated by strand and chromosome end, with full length Nanopore read sequences\n",
    "with open(outlog_fh, \"w\") as log_out:#write statistic result\n",
    "    log_out.write(\"Stage\\tChromosome\\tEnd\\tValue\\tAttribute\\n\")#Telomere reads count\n",
    "    for stage in stg:\n",
    "        for chrom in chromosome:\n",
    "            for position in chrend:\n",
    "                #read in files path\n",
    "                raw_fasta = str(\"fasta/%s_%s%s.fa\" % (stage, chrom, position))#input fasta\n",
    "                blastn_file = str(\"blastn/%s_%s%s.tsv\" % (stage, chrom, position))#input blastn result\n",
    "                read_list_file = str(\"read_list_blastn/%s_%s%s_list\" % (stage, chrom, position))#input readname list\n",
    "                #output file path\n",
    "                output_fh = str(\"trimming_result/telomere_only/%s_%s%s.fa\" % (stage, chrom, position)) #output file path\n",
    "                forward_output_fh = str(\"trimming_result/strand/%s_%s%s_F.fa\" % (stage, chrom, position)) #output file path\n",
    "                reverse_output_fh = str(\"trimming_result/strand/%s_%s%s_R.fa\" % (stage, chrom, position)) #output file path\n",
    "                #read input fasta and blastn result to convert sequences from central chromosome to end\n",
    "                read_dict = SeqIO.index(raw_fasta, \"fasta\")#index extracted read sequence from BAM file\n",
    "                blastn_info = pd.read_csv(blastn_file, sep = \"\\t\",header=None,index_col=False)#pandas read blastn result \n",
    "                blastn_info.columns = [\"query\", \"subject\", \"identity\", \"length\", \"mis\",\"gap\",\"qstart\",\"qend\",\"sstart\",\"send\",\"evalue\",\"bit\",\"strand\"]#assgin column names\n",
    "                forward_counter = 0\n",
    "                forward_total_length = 0\n",
    "                reverse_counter = 0\n",
    "                reverse_total_length = 0\n",
    "                short_counter = 0\n",
    "                non_end_counter = 0\n",
    "                chimeric_counter = 0\n",
    "                self_duplication_counter = 0\n",
    "                read_total_length = 0\n",
    "                telomere_total_length = 0\n",
    "                with open(output_fh , \"w\") as out_w ,open(read_list_file , \"r\") as read_list, \\\n",
    "                open(forward_output_fh , \"w\") as forward_out_w , open(reverse_output_fh , \"w\") as reverse_out_w:\n",
    "                    for line in read_list:\n",
    "                        readName = line.rstrip()\n",
    "                        read_seq = str(read_dict[readName].seq)\n",
    "                        trimSeq(blastn_filter(chrom, position, readName))#write trimmed sequences to file\n",
    "                out_w.close()\n",
    "                forward_out_w.close()\n",
    "                reverse_out_w.close()\n",
    "                read_list.close()\n",
    "                #write statistic results to trim_fasta/log/ by chromosome end position\n",
    "                write_log(stage, chrom, position)\n",
    "log_out.close()\n",
    "print(\"\\njob finished!\")            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn\n",
      "blastn_selected\n",
      "chr_ref\n",
      "fasta\n",
      "list_filter\n",
      "list_raw\n",
      "read_list_blastn\n",
      "ref\n",
      "trf_data\n",
      "trim_fasta\n",
      "trimmed\n",
      "trimming_result\n"
     ]
    }
   ],
   "source": [
    "#analyze sequence length with seqkit, telomere repeat with Tidehunter, and get the trimmed telomere sequences read name\n",
    "#you can run it in terminal\n",
    "%%bash\n",
    "mkdir -p trimming_result/tidehunter/\n",
    "mkdir -p trimming_result/tidehunter/P8p4k3/\n",
    "mkdir -p trimming_result/tidehunter/stats/\n",
    "mkdir -p trimming_result/ONT_reads/\n",
    "mkdir -p trimming_result/log/\n",
    "stage=(\"N2EM\" \"N2L1\" \"N2YA\")\n",
    "chrom=(\"I\" \"II\" \"III\" \"IV\" \"V\" \"X\")\n",
    "position=(\"L\" \"R\")\n",
    "strand=(\"F\" \"R\")\n",
    "#run command in bash\n",
    "for i in ${stage[@]}\n",
    "do\n",
    "for j in ${chrom[@]}\n",
    "do\n",
    "for k in ${position[@]}\n",
    "do\n",
    "for l in  ${strand[@]}\n",
    "do\n",
    "samtools faidx trimming_result/telomere_only/${i}_${j}${k}.fa\n",
    "samtools faidx trimming_result/strand/${i}_${j}${k}_${l}.fa\n",
    "seqkit stats -a trimming_result/telomere_only/${i}_${j}${k}.fa -T|tail -1 >> trimming_result/log/N2_telomere_stats.tsv\n",
    "TideHunter -f 2 -P 8 -p 4 -k 3 trimming_result/strand/${i}_${j}${k}_${l}.fa > trimming_result/tidehunter/${i}_${j}${k}_${l}_P8p4k3.tsv\n",
    "seqkit seq trimming_result/telomere_only/${i}_${j}${k}.fa -n -i> trimming_result/telomere_only/${i}_${j}${k}_list\n",
    "seqkit seq trimming_result/strand/${i}_${j}${k}_${l}.fa -n -i> trimming_result/ONT_reads/${i}_${j}${k}_${l}_list\n",
    "seqtk subseq fasta/${i}_${j}${k}.fa trimming_result/ONT_reads/${i}_${j}${k}_${l}_list > trimming_result/ONT_reads/${i}_${j}${k}_${l}.fa\n",
    "samtools faidx trimming_result/ONT_reads/${i}_${j}${k}_${l}.fa\n",
    "done\n",
    "done\n",
    "done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TTAGGC', 'TAGGCT', 'AGGCTT', 'GGCTTA', 'GCTTAG', 'CTTAGG']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#generate a canonical telomere repeat and loop list\n",
    "canonical = \"TTAGGC\"#canonical telomere repeat in C. elegans\n",
    "canonical_list = enumerateRepeat(canonical)#looping telomere sequence\n",
    "canonical_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare directory to save results\n",
    "!mkdir -p trimming_result/tidehunter/stats/coverage/\n",
    "!mkdir -p trimming_result/tidehunter/stats/unsort/\n",
    "!mkdir -p trimming_result/tidehunter/stats/sorted/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N2YA XR R 52th 5-mer motif: TTCAGTTTTGA\n",
      "Merging repeat finished\n"
     ]
    }
   ],
   "source": [
    "#make telomere repeat statistic file and coverage file\n",
    "for stage in stg:\n",
    "    for chrom in chromosome:\n",
    "        for position in chrend:\n",
    "            for direction in strand:\n",
    "                #print(\"\\n=========={} {}{} {} strand start==========\".format(stage, chrom, position, direction))\n",
    "                #read in files path\n",
    "                tide_file = str(\"trimming_result/tidehunter/%s_%s%s_%s_P8p4k3.tsv\" % (stage, chrom, position, direction))#input Tidehunter tabular results\n",
    "                trimmed_read_list_file = str(\"trimming_result/ONT_reads/%s_%s%s_%s_list\" % (stage, chrom, position, direction))#input readname list\n",
    "                #output file path\n",
    "                output_log_fh = str(\"trimming_result/tidehunter/stats/unsort/%s_%s%s_%s_stats.tsv\" % (stage, chrom, position, direction)) #output file path\n",
    "                output_table_fh = str(\"trimming_result/tidehunter/stats/coverage/%s_%s%s_%s_Coverage.tsv\" % (stage, chrom, position, direction)) #output coverage file path\n",
    "                #input Tidehunter tabular result as tide_info with pandas\n",
    "                tide_info = pd.read_csv(tide_file, sep = \"\\t\",header=None,index_col=False)#pandas read blastn result \n",
    "                tide_info.columns = [\"readName\", \"consN\", \"readLen\", \"start\", \"end\",\"consLen\",\"copyNum\",\"fullLen\",\"subPos\",\"consensus\"]#assgin 10 column names to Tidehunter table\n",
    "                Coverage = pd.DataFrame({\"TTAGGC\": [0] * 10000})#make a single column coverage df for canonical repeat\n",
    "                new_motif_counter = 0#newly found repeat counter\n",
    "                canonical_copyNum = 0#copy number caculater\n",
    "                with open(trimmed_read_list_file,\"r\") as read_list, open(output_log_fh, \"w\") as out_log:\n",
    "                    for line in read_list:#readIn read name from line\n",
    "                        readName = line.rstrip()#get readName from line in read name list file\n",
    "                        df_read_repeat = tide_filter(readName)#select rows with readName\n",
    "                        #looping rows for read\n",
    "                        for idx in range(len(df_read_repeat)):\n",
    "                            working_motif = df_read_repeat.iloc[idx].consensus#assign name for the consensus in the row\n",
    "                            if working_motif in canonical_list:#if consensus is canonical repeat sequence in the canonical list\n",
    "                                canonical_copyNum += df_read_repeat.iloc[idx].copyNum\n",
    "                                for i in range(df_read_repeat.iloc[idx].start - 1, df_read_repeat.iloc[idx].end):\n",
    "                                    Coverage.TTAGGC[i] += 1#add data to table\n",
    "                            #if consensus is not a canonical telomere repeat\n",
    "                            elif new_motif_counter == 0:#if working consensus is the first found repeat\n",
    "                                new_motif_counter += 1#new repeat found +1\n",
    "                                tmp_name = \"new_repeat\" + str(new_motif_counter)#assign name\n",
    "                                sys.stdout.write('\\r'+str(\"%s %s%s %s %dst %d-mer motif: %s\" %(stage, chrom, position, direction, new_motif_counter, len(working_motif), working_motif)))#print newly identified repeat\n",
    "                                vars()[tmp_name] = enumerateRepeat(working_motif)#generate a list for the first found new repeat\n",
    "                                vars()[tmp_name + \"_repeat\"] = working_motif#assign consensus sequences to a new repeat \n",
    "                                Coverage[tmp_name] = [0] * 10000#add a empty column to table for the new repeat\n",
    "                                repeat_col = \"Coverage.\" + tmp_name#assign name\n",
    "                                for i in range(df_read_repeat.iloc[idx].start - 1, df_read_repeat.iloc[idx].end):\n",
    "                                    eval(repeat_col)[i] += 1#add data of new repeat to table\n",
    "                            else:\n",
    "                                observed = \"False\"\n",
    "                                for k in range(new_motif_counter):#search newly found repeats\n",
    "                                    tmp_name = \"new_repeat\" + str(k +1)#assign name\n",
    "                                    if working_motif in vars()[tmp_name]:#if repeat has been found\n",
    "                                        repeat_col = \"Coverage.\" + tmp_name#assign name\n",
    "                                        vars()[tmp_name + \"_repeat\"] = working_motif#assign consensus sequences to new repeat \n",
    "                                        observed = \"True\"\n",
    "                                        for i in range(df_read_repeat.iloc[idx].start - 1, df_read_repeat.iloc[idx].end):\n",
    "                                            eval(repeat_col)[i] += 1#add data to table\n",
    "                                        break\n",
    "                                if observed == \"False\": #if working consensus is a new repeat\n",
    "                                    new_motif_counter += 1#new repeat found +1\n",
    "                                    tmp_name = \"new_repeat\" + str(new_motif_counter)#assign name\n",
    "                                    sys.stdout.write('\\r'+str(\"%s %s%s %s %dth %d-mer motif: %s\" %(stage, chrom, position, direction, new_motif_counter, len(working_motif), working_motif)))#print newly identified repeat\n",
    "                                    vars()[tmp_name] = enumerateRepeat(working_motif)#generate a list for the newly found repeat\n",
    "                                    vars()[tmp_name + \"_repeat\"] = working_motif#assign consensus sequences to new repeat \n",
    "                                    Coverage[tmp_name] = [0] * 10000#add a empty column to table\n",
    "                                    repeat_col = \"Coverage.\" + tmp_name\n",
    "                                    for i in range(df_read_repeat.iloc[idx].start - 1, df_read_repeat.iloc[idx].end):\n",
    "                                        eval(repeat_col)[i] += 1\n",
    "                    out_log.write(\"RepeatSeq\\tConsensus\\tCopyNum\\n\")\n",
    "                    out_log.write(\"canonical\\t{}\\t{:.1f}\\n\".format(canonical, canonical_copyNum))\n",
    "                    if new_motif_counter > 0:\n",
    "                        for k in range(new_motif_counter):\n",
    "                            repeat_name =  \"new_repeat\" + str(k+1)\n",
    "                            out_log.write(\"{}\\t{}\\t{:.1f}\\n\".format(repeat_name, vars()[repeat_name + \"_repeat\"], eval(\"Coverage.\" + repeat_name + \".sum()\") / 6))\n",
    "                read_list.close()\n",
    "                out_log.close()\n",
    "                Coverage.to_csv(output_table_fh, index=False, header=True, sep=\"\\t\")\n",
    "print(\"\\nMerging repeat finished\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=====Job finished=====\n"
     ]
    }
   ],
   "source": [
    "#sort telomere repeat stats file\n",
    "for stage in stg:\n",
    "    for chrom in chromosome:\n",
    "        for position in chrend:\n",
    "            for direction in strand:\n",
    "                #read in files path\n",
    "                unsort_file = str(\"trimming_result/tidehunter/stats/unsort/%s_%s%s_%s_stats.tsv\" % (stage, chrom, position, direction))#input table\n",
    "                sort_fh = str(\"trimming_result/tidehunter/stats/sorted/%s_%s%s_%s_stats_sorted.tsv\" % (stage, chrom, position, direction)) #output sorted table\n",
    "                #input Tidehunter tabular result as tide_info with pandas\n",
    "                unsort_info = pd.read_csv(unsort_file, sep = \"\\t\",header= 0,index_col=False)#pandas read table \n",
    "                sort_info = unsort_info.sort_values(by =\"CopyNum\" , ascending = False)\n",
    "                sort_info.to_csv(sort_fh, index = False, header = True, sep=\"\\t\")\n",
    "                #sys.stdout.write('\\r'+str(\"%s %s%s %s finished\" %(stage, chrom, position, direction)))\n",
    "print(\"=====Job finished=====\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pipeline end up here.\n",
    "\n",
    "Following data analysis is performed in **Rstudio** with R v.3.6.1 or above\n",
    "\n",
    "Required R packages:\n",
    "\n",
    "**ggplot2**\n",
    "\n",
    "**reshape2**\n",
    "\n",
    "**grid**\n",
    "\n",
    "**gridExtra**\n",
    "\n",
    "**dplyr**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
